provider/aws: aws_db_instance: make some fields optional

koendc opened the pull request:
When you create an RDS instance from a snapshot, the parameters username, password,
engine, engine_version and allocated_storage are not mandatory. 
The value can be taken from the snapshot, so no need to supply them.

apparentlymart commented on website/source/docs/providers/aws/r/db_instance.html.markdown:
```diff
@@ -34,9 +34,9 @@ the [AWS official documentation](http://docs.aws.amazon.com/AmazonRDS/latest/Com
-* `allocated_storage` - (Required) The allocated storage in gigabytes.
-* `engine` - (Required) The database engine to use.
-* `engine_version` - (Required) The engine version to use.
+* `allocated_storage` - (Optional when a `snapshot_identifier` is provided) The allocated storage in gigabytes.
```
This is totally a nitpick, but I think I'd phrase this as "Required unless a `snapshot_identifier` is provided" to describe the situation with slightly more precision.

apparentlymart commented:
Do you think it's worth having some validation code called from the ``Create`` and ``Update`` functions to enforce the rules described in the docs, or does the AWS API produce a reasonable error message itself?

koendc commented:
I will try to come up with some validation code.

koendc commented:
I have updated the documentation: the fields are also not used when creating read replica's.
I have added validation of the required fields in the `Create` function. I don't think it makes sense for the `Update` function to have the validation.

catsby commented on builtin/providers/aws/resource_aws_db_instance_test.go:
```diff
@@ -71,32 +72,55 @@ func testAccCheckAWSDBInstanceDestroy(s *terraform.State) error {
-		if rs.Type != "aws_db_instance" {
-			continue
-		}
+		if rs.Type == "aws_db_instance" {
```
can you help me understand why these changes were made? Why are we now checking `aws_db_subnet_group` ?

catsby commented on website/source/docs/providers/aws/r/db_instance.html.markdown:
```diff
@@ -34,8 +34,8 @@ the [AWS official documentation](http://docs.aws.amazon.com/AmazonRDS/latest/Com
 
-* `allocated_storage` - (Required) The allocated storage in gigabytes.
-* `engine` - (Required) The database engine to use.
+* `allocated_storage` - (Required unless a `snapshot_identifier` is provided) The allocated storage in gigabytes.
```
We should also note now that they are not required for read replicas, correct?

catsby commented:
Hello –

Thanks for the patch! I apologize for the silence. I've made a few comments that I'd like to hear about before merging, but otherwise things look good. Thanks!

koendc commented:
Hi @catsby, something went wrong with that one commit you were commenting on. I removed the unintended changes to the tests.
As you suggested, I added a read replicas to the documentation.

catsby commented:
Looks good, thanks!
